DatabaseServer/moxa_data.py

```diff
@@ -126,10 +131,30 @@ def _get_mappings_str(self):
         newmap = dict()
         for hostname, mappings in self._mappings[1].items():
             ip_addr = self._mappings[0][hostname]
+            mibmap = walk(ip_addr, '1.3.6.1.2.1', SYSTEM_MIBS + PORT_MIBS)
```

Comment:
The `_get_mappings_str` function is called indirectly from https://github.com/ISISComputingGroup/EPICS-inst_servers/blob/1dae948915e00bf6c430e02091b29108996e1f00/DatabaseServer/database_server.py#L122 when a read of the process variable is requested. Due to now polling moxas via `walk` this function can take a while to execute, and this can affect the database server process performance. To improve this we could move the snmp walking to a separate thread so `_get_mapping_str` just returns the most recent cached information.

Reply:
Ideally if a Moxa is switched off it shouldn't be part of the dictionary which it is iterating on in the _get_mappings_str method. Is there a requirement to show non-functional MOXA devices on the IBEX GUI?

Reply:
when a moxa is added to an instrument its details gets written into the windows registry and will be there until removed by us, the initial dictionary contains these details which bear no relation to whether the moxa still even exists.  If a moxa is detailed in the windows registry but not contactable, we don't know if that is a problem (it is really needed but has failed) or not (it is either an old registry entry, or it belongs to a piece of mobile equipment whose temporary absence is expected). So we need to show non-contactable devices and let the user decide, if it really has been removed hopefully they will contact us to "tidy up their display". So we need to indicate non contactable moxas and also make sure they do not affect the performance of the system. Rather than a thread, a simple initial "ping" might be good enough, but we will still need to decide on a timeout for the ping which may not in the end save time. A separate data collecting thread is probably the most robust, or a possibly a thread that just does a ping and so keeps a list of which moxas are on the network and so the existing approach just checks a subset of hosts with snmp. Are there any other approaches you can think of?

Reply:
Yep, I saw that the code reads data from the registry and so it would retain information for the non-functional ones as well. That's why I asked whether there is such a requirement. Thanks for the clarification. I now understand why this is there.
For the SNMP walk I think there is a way to set a timeout and also retry count. I can try adding a timeout and see if it works. It could be something like 10 seconds.